Contain memory operands under casts

SingleAccretion opened the pull request:
Recognize legal patterns in lowering and then fold the sign/zero-extension into an appropriate load at codegen time.

The change implements full support on XARCH, partial support on ARM/64, and punts LA. A follow-up issue will be filed to track the remaining work. The ARM/64 commit contains some notes on what is required for said support on our load/store architectures.

The [diffs](https://dev.azure.com/dnceng/public/_build/results?buildId=1899864&view=results) are nice and simple:
```diff
-            ldr     lr, [sp+0x34]      // [V07 arg7]
-            uxtb    lr, lr
+            ldrb    lr, [sp+0x34]      // [V07 arg7]
```
Regressions are RA/alignment.

SingleAccretion committed:
Cast descriptor support

SingleAccretion committed:
XARCH support

SingleAccretion committed:
ARM/ARM64 support

TODO: consider using a dedicated IND_EXT oper for ARM/ARM64 instead of containment.
This would allow us to cleany handle all indirections. It would not mean we'd give
up on the casts containment, as we'd still need to handle the "reg optional" case.

IND_EXT will be much like an ordinary IND, but have a "source" and "target" types.
The "target" type would always be int/long, while "source" could be of any integral
type.

This design would be a bit more natural, and nicely separable from casts. However,
the main problem with the current state of things, apart from the fact codegen of
indirections is tied strongly to "GenTreeIndir", is the fact that changing type of
the load can invalidate LEA containment. One would think this is solvable with some
tricks, like re-running containment analysis on an indirection after processing the
cast, but ARM64 codegen doesn't support uncontained LEAs in some cases.

A possible solution to that problem is uncontaining the whole address tree. That
would be messy, but ought to work. An additional complication is that these trees
can contain a lot of contained operands as part of ADDEX and BFIZ, so what would
have to be done first is the making of these into proper EXOPs.

In any case, this is all future work.

SingleAccretion committed:
Add GenTreeCast::IsZeroExtending

SingleAccretion commented:
@dotnet/jit-contrib

TIHan commented on src/coreclr/jit/codegenarmarch.cpp:
```diff
@@ -3946,15 +3946,70 @@ void CodeGen::genIntToIntCast(GenTreeCast* cast)
+            emit->emitIns_Mov(ins, EA_ATTR(insSize), dstReg, srcReg, /* canSkip */ false);
+        }
+        else
+        {
```
I'm not sure what this block is doing, is it fair to add a few comments describing it?

SingleAccretion replied:
Would below help?
```
// The "used from memory" case. On ArmArch casts are the only nodes which can have
// contained memory operands, so we have to handle all possible sources "manually".
assert(src->isUsedFromMemory());
```
This `else` is just handling the "used from memory" cases, of which there are three: "used from a spill temp" (the source was non-local node for which LSRA spilled the register), "contained in-memory local", and finally an explicit indir.

One can find a lot of code like this on XArch, e. g. in `CodeGen::genOperandDesc`, `emitInsBinary`, etc.

TIHan replied:
Yea that would help. Thank you!

TIHan reviewed (approved):
The diffs look really good.

I tried doing something similar here: https://github.com/dotnet/runtime/pull/70756 - but I didn't get to work on it longer.

SingleAccretion committed:
Cleanup, fix & a comment

tannergooding commented on src/coreclr/jit/codegenxarch.cpp:
```diff
@@ -6804,33 +6802,51 @@ void CodeGen::genIntToIntCast(GenTreeCast* cast)
             break;
 #ifdef TARGET_64BIT
         case GenIntCastDesc::ZERO_EXTEND_INT:
+        case GenIntCastDesc::LOAD_ZERO_EXTEND_INT:
```
Are there going to be any issues with `Load` variants and `canSkip` or will this always be "false" because `srcReg` is `REG_NA` or something else?

SingleAccretion replied:
> will this always be "false" because srcReg is REG_NA or something else?

That's right, because of this code in `AreUpper32BitsZero`:
```cpp
// Bail if not writing to the right register
if (id->idReg1() != reg)
{
    return false;
}
```
Additionally, `canSkip` is not used on the "load" path so it does not really matter.

SingleAccretion commented:
I think it would be good to run the fuzzers and [libraries] stress on this change.

EgorBo commented:
/azp run Fuzzlyn, Antigen, runtime-coreclr jitstress, runtime-coreclr libraries-jitstress, runtime-coreclr gcstress0x3-gcstress0xc, runtime-coreclr jitstressregs

SingleAccretion committed:
Fix an issue found by the fuzzers

In CAST<short>(IND<byte>(...)), "m_extendSrcSize" must be "1".

Modulo the above, stress runs looked clean.

jakobbotsch reviewed (approved):
It's curious to me that there are so many x86 improvements. Do we not already removed most of these unnecessary casts there by retyping loads in morph?

SingleAccretion commented:
> It's curious to me that there are so many x86 improvements. Do we not already removed most of these unnecessary casts there by retyping loads in morph?

I think most of these come from normalize-on-load stack parameters, which we do not retype.

jakobbotsch commented:
> I think most of these come from normalize-on-load stack parameters, which we do not retype.

Doesn't morph introduce this IR shape itself in `fgMorphLocalVar`? So doing that is a large regression?

SingleAccretion commented:
> Doesn't morph introduce this IR shape itself in `fgMorphLocalVar`? So doing that is a large regression?

It of course does, and it is, but it is necessary for in-register parameters (and on-stack ones as well, though that should be relatively easy to fix by using extending loads in `genEnregisterIncomingStackArgs`).

Morph's logic there is questionable in more than one way. If we know the local will be DNER at that point, we could retype it to its small type instead of introducing a cast. However, this turns out to sometimes not be a win because we don't CSE locals, but do CSE casts.

jakobbotsch commented:
Makes sense about CSE. The way `fgMorphLocalVar` worked always seemed odd to me.

Anyway, thanks for the contribution as usual.

DrewScoggins commented:
Improvements on Win-x86: https://github.com/dotnet/perf-autofiling-issues/issues/8170

AndyAyersMS commented:
Possible regression on arm64: https://github.com/dotnet/perf-autofiling-issues/issues/8221
